pkg/virt-controller/watch/vm.go

```diff
@@ -271,7 +271,110 @@ func (c *VMController) execute(key string) error {
 		return err
 	}
 
-	return createErr
+	if createErr != nil {
+		return createErr
+	}
+
+	err = handleVMRenaming(vm, c, logger)
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func handleVMRenaming(vm *virtv1.VirtualMachine, c *VMController, logger *log.FilteredLogger) error {
+	// Check if renaming annotation exists
+	if vm.ObjectMeta.Annotations == nil {
+		return nil
+	}
+
+	renameTo, hasRenameToAnnotation := vm.ObjectMeta.Annotations[virtv1.RenameToAnnotation]
+
+	// Handle source VM
+	if hasRenameToAnnotation {
+		// If a VM with the new name exists, remove annotation
+		_, err := c.clientset.VirtualMachine(vm.Namespace).Get(renameTo, &v1.GetOptions{})
+
+		if err != nil {
+			if !strings.Contains(err.Error(), "not found") {
+				errMsg :=
+					fmt.Sprintf("can't determine existence of newVM: %s while processing rename of oldVM: %s",
+						renameTo, vm.Name)
+				logger.Reason(err).Error(errMsg)
+				return err
+			} else {
+				// New vm is not found, create it
+				newVM := &virtv1.VirtualMachine{
+					TypeMeta: vm.TypeMeta,
+					ObjectMeta: v1.ObjectMeta{
+						Name:      renameTo,
+						Namespace: vm.Namespace,
+						Labels:    vm.Labels,
+						Annotations: map[string]string{
+							virtv1.RenameFromAnnotation: vm.Name,
+						},
+						// Finalizers: vm.Finalizers,
+					},
+					Spec: vm.Spec,
+				}
+
+				_, err := c.clientset.VirtualMachine(vm.Namespace).Create(newVM)
+
+				if err != nil {
+					errMsg := fmt.Sprintf("Could not create a new vm during the rename process of: %s", vm.Name)
+					logger.Reason(err).Error(errMsg)
+					return err
+				}
+			}
+		}
+
+		// New VM exists, remove renameTo annotation
+		patchString := fmt.Sprintf("{\"metadata\":{\"annotations\":{\"%s\":null}}}",
+			virtv1.RenameToAnnotation)
+		_, err = c.clientset.VirtualMachine(vm.Namespace).Patch(vm.Name, types.MergePatchType, []byte(patchString))
+
+		if err != nil {
+			logger.Reason(err).Error(fmt.Sprintf("Failed removing 'renameTo' annotation from vm: %s", vm.Name))
+			return err
+		}
+
+		return nil
+	}
+
+	renameFrom, hasRenameFromAnnotation := vm.ObjectMeta.Annotations[virtv1.RenameFromAnnotation]
```

Comment:
I think that the flow here can go into a different function.. maybe `removeRenamedVM`?

Reply:
I don't know, I think this function is pretty straight-forward in the way it handles both source/target VMs and it's not too big